src/aliased_buffer.h

```diff
@@ -0,0 +1,199 @@
+    const v8::HandleScope handle_scope(isolate_);
+    // validate that the byte_offset is aligned with sizeof(NativeT)
+    CHECK_EQ(byte_offset & (sizeof(NativeT) - 1), 0);
+    // validate this fits inside the backing buffer
+    CHECK_LE(sizeof(NativeT) * count,
+        backing_buffer.GetArrayBuffer()->ByteLength() - byte_offset);
+
+    buffer_ = reinterpret_cast<NativeT*>(
+        const_cast<uint8_t*>(backing_buffer.GetNativeBuffer() + byte_offset));
+
+    v8::Local<v8::ArrayBuffer> ab = backing_buffer.GetArrayBuffer();
+    v8::Local<V8T> js_array = V8T::New(ab, byte_offset, count);
+    js_array_ = v8::Global<V8T>(isolate, js_array);
+  }
+
+  AliasedBuffer(const AliasedBuffer& that)
+      : isolate_(that.isolate_),
+        count_(that.count_),
+        byte_offset_(that.byte_offset_),
+        buffer_(that.buffer_),
+        free_buffer_(false) {
+    js_array_ = v8::Global<V8T>(that.isolate_, that.GetJSArray());
+  }
+
+  ~AliasedBuffer() {
+    if (free_buffer_ && buffer_ != nullptr) {
+      free(buffer_);
+    }
+    js_array_.Reset();
+  }
+
+  /**
+   * Helper class that is returned from operator[] to support assignment into
+   * a specified location.
+   */
+  class Reference {
+   public:
+    Reference(AliasedBuffer<NativeT, V8T>* aliased_buffer, size_t index)
+        : aliased_buffer_(aliased_buffer),
+          index_(index) {
+    }
+
+    Reference(const Reference& that)
+        : aliased_buffer_(that.aliased_buffer_),
+          index_(that.index_) {
+    }
+
+    inline Reference& operator=(const NativeT &val) {
+      aliased_buffer_->SetValue(index_, val);
+      return *this;
+    }
+
+    operator NativeT() const {
+      return aliased_buffer_->GetValue(index_);
+    }
+
+   private:
+    AliasedBuffer<NativeT, V8T>* aliased_buffer_;
+    size_t index_;
+  };
+
+  /**
+   *  Get the underlying v8 TypedArray overlayed on top of the native buffer
+   */
+  v8::Local<V8T> GetJSArray() const {
+    return js_array_.Get(isolate_);
+  }
+
+  /**
+  *  Get the underlying v8::ArrayBuffer underlying the TypedArray and
+  *  overlaying the native buffer
+  */
+  v8::Local<v8::ArrayBuffer> GetArrayBuffer() const {
+    return GetJSArray()->Buffer();
+  }
+
+  /**
+   *  Get the underlying native buffer. Note that all reads/writes should occur
+   *  through the GetValue/SetValue/operator[] methods
+   */
+  inline const NativeT* GetNativeBuffer() const {
+    return buffer_;
+  }
+
+  /**
+   *  Synonym for GetBuffer()
+   */
+  inline const NativeT* operator * () const {
+    return GetNativeBuffer();
+  }
+
+  /**
+   *  Set position index to given value.
+   */
+  inline void SetValue(const size_t index, NativeT value) {
+    #if defined(DEBUG) && DEBUG
+      CHECK_LT(index, count_);
+    #endif
+    buffer_[index] = value;
+  }
+
+  /**
+   *  Get value at position index
+   */
+  inline const NativeT GetValue(const size_t index) const {
+    #if defined(DEBUG) && DEBUG
+      CHECK_LT(index, count_);
+    #endif
```

Comment:
nit (non-blocking): more common indentation is to have no whitespace before the `#if`.

Reply:
fixed